x/distribution/keeper/delegation.go

```diff
@@ -53,28 +55,44 @@ func (k Keeper) calculateDelegationRewardsBetween(ctx sdk.Context, val sdk.Valid
 func (k Keeper) calculateDelegationRewards(ctx sdk.Context, val sdk.Validator, del sdk.Delegation, endingPeriod uint64) (rewards sdk.DecCoins) {
 	// fetch starting info for delegation
 	startingInfo := k.GetDelegatorStartingInfo(ctx, del.GetValidatorAddr(), del.GetDelegatorAddr())
+
+	if startingInfo.Height == uint64(ctx.BlockHeight()) {
+		// started this height, no rewards yet
+		return
+	}
+
 	startingPeriod := startingInfo.PreviousPeriod
 	stake := startingInfo.Stake
 
 	// iterate through slashes and withdraw with calculated staking for sub-intervals
 	// these offsets are dependent on *when* slashes happen - namely, in BeginBlock, after rewards are allocated...
-	// ... so we don't reduce stake for slashes which happened in the *first* block, because the delegation wouldn't have existed
-	startingHeight := startingInfo.Height + 1
-	// ... or slashes which happened in *this* block, since they would have happened after reward allocation
-	endingHeight := uint64(ctx.BlockHeight()) - 1
+	// slashes which happened in the first block would have been before this delegation existed,
+	// UNLESS they were slashes of a redelegation to this validator which was itself slashed earlier in the same BeginBlock
+	startingHeight := startingInfo.Height
+	// slashes this block happened after reward allocation, but we have to account for them for the stake sanity check
```

Comment:
By "stake sanity check", is this referring to the sanity check below in the same function?  Elaborate would help, e.g. "stake sanity check **below**", and "// a stake sanity check"

Reply:
Added.